service/src/main/java/bio/terra/pipelines/service/PipelineRunsService.java

```diff
@@ -0,0 +1,185 @@
+            .addParameter(RunImputationJobFlightMapKeys.PIPELINE_ID, pipeline.getId())
+            .addParameter(
+                RunImputationJobFlightMapKeys.PIPELINE_INPUT_DEFINITIONS,
+                pipeline.getPipelineInputDefinitions())
+            .addParameter(
+                RunImputationJobFlightMapKeys.USER_PROVIDED_PIPELINE_INPUTS, userProvidedInputs)
+            .addParameter(
+                RunImputationJobFlightMapKeys.CONTROL_WORKSPACE_ID,
+                pipeline.getWorkspaceId().toString())
+            .addParameter(
+                RunImputationJobFlightMapKeys.WDL_METHOD_NAME, pipeline.getWdlMethodName())
+            .addParameter(JobMapKeys.RESULT_PATH.getKeyName(), resultPath);
+
+    jobBuilder.submit();
+
+    logger.info("Created {} pipelineRun with jobId {}", pipelineName, jobId);
+
+    return pipelineRun;
+  }
+
+  public PipelineRun writePipelineRunToDb(
+      UUID jobUuid,
+      String userId,
+      Long pipelineId,
+      CommonPipelineRunStatusEnum status,
+      String description,
+      String resultUrl,
+      Map<String, Object> pipelineInputs) {
+
+    // write pipelineRun to database
+    PipelineRun pipelineRun =
+        new PipelineRun(jobUuid, userId, pipelineId, status.toString(), description, resultUrl);
+    PipelineRun createdPipelineRun = writePipelineRunToDbThrowsDuplicateException(pipelineRun);
+
+    String pipelineInputsAsString;
+    try {
+      // do this to write the pipeline inputs without writing the class name
+      pipelineInputsAsString = objectMapper.writeValueAsString(pipelineInputs);
+    } catch (JsonProcessingException e) {
+      // this should never happen
+      throw new InternalServerErrorException("Internal error processing pipeline inputs", e);
+    }
+
+    // save related pipeline inputs
+    PipelineInput pipelineInput = new PipelineInput();
+    pipelineInput.setJobId(createdPipelineRun.getId());
+    pipelineInput.setInputs(pipelineInputsAsString);
+    pipelineInputsRepository.save(pipelineInput);
+
+    return createdPipelineRun;
+  }
+
+  protected PipelineRun writePipelineRunToDbThrowsDuplicateException(PipelineRun pipelineRun)
+      throws DuplicateObjectException {
+    try {
+      pipelineRunsRepository.save(pipelineRun);
+      logger.info("pipelineRun saved for jobId: {}", pipelineRun.getJobId());
+    } catch (DataIntegrityViolationException e) {
+      if (e.getCause() instanceof ConstraintViolationException c
+          && c.getConstraintName().contains("jobId_unique")) {
+        throw new DuplicateObjectException(
+            String.format("Duplicate jobId %s found", pipelineRun.getJobId()));
+      }
+      throw e;
+    }
+
+    return pipelineRun;
+  }
+
+  public PipelineRun getPipelineRun(UUID jobId, String userId) {
+    return pipelineRunsRepository.findByJobIdAndUserId(jobId, userId).orElse(null);
+  }
+
+  public PipelineRun markPipelineRunSuccess(UUID jobId, String userId) {
```

Comment:
there's currently no logic here or anywhere to assert that the run actually WAS a success. my thinking was that if we get to the flight step that calls this method, it will only be in the case that the run has succeeded; any error will stop the flight before it gets there. I'm not sure whether to:
1) assume that's the case
2) check here that that's the case and fail if not
3) write the status to SUCCEEDED here regardless of what the previous status was
4) figure out a way to test this

Reply:
I think that if the logic is, we got to the final step and therefore the flight was a success, then we should be marking the status in our db as SUCCESS here.

Reply:
discussed offline - added docstring for now